MSNeventListener.java

```diff
@@ -0,0 +1,192 @@
+        }
+
+        public int dwData;
+        public long cbData;
+        public Pointer lpData;
+
+        protected List<String> getFieldOrder() {
+            return Arrays.asList(new String[] { "dwData", "cbData", "lpData" });
+        }
+    }
+
+    /**
+     * The runnable MsnMessagePump handles the messages normally received by the MSN messenger.
+     */
+    public Runnable MsnMessagePump = new Runnable() {
+        @Override
+        public void run() {
+            WNDCLASSEX msnWndClass = new WNDCLASSEX();
+            msnWndClass.lpszClassName = lpClassName;
+            msnWndClass.lpfnWndProc = WNDPROC;
+            if (MyUser32.INSTANCE.RegisterClassEx(msnWndClass).intValue() > 0) {
+
+                /* Create a native window */
+                HWND hMsnWnd = MyUser32.INSTANCE.CreateWindowEx(0, lpClassName, "", 0,
+                            0, 0, 0, 0, null, null, null, null);
+
+                /* Register the callback */
+                try {
+                    // Use SetWindowLongPtr if available (64-bit safe)
+                    MyUser32.INSTANCE.SetWindowLongPtr(hMsnWnd, GWL_WNDPROC, WNDPROC);
+                } catch(UnsatisfiedLinkError e) {
+                    // Use SetWindowLong if SetWindowLongPtr isn't available
+                    MyUser32.INSTANCE.SetWindowLong(hMsnWnd, GWL_WNDPROC, WNDPROC);
+                }
+
+                /* Handle events until the window is destroyed */
+                MSG msg = new MSG();
+                msg.clear();
+                while(MyUser32.INSTANCE.GetMessage(msg, hMsnWnd, 0, 0) > 0) {
+                    MyUser32.INSTANCE.TranslateMessage(msg);                                 
+                    MyUser32.INSTANCE.DispatchMessage(msg);
+                }
+            }
+        }
+    };
+    
+    /**
+     * The runnable OnMSNStringChanged encrypts and returns received message data.
+     */
+    public Runnable OnMSNStringChanged = new Runnable() {
+  
+        @Override
+        public void run() {
+            if (_msnArray) {
+				/*
+				 * Do stuff with the message data in msnArray here.
+				 * Ex. label and put values into a new HashMap.
+				 */
+            }
+        }
+    };
+    /**
+     * The window procedure WNDPROC overrides the default callback method with a customized one.
+     * @returns LRESULT
+     */
+    WindowProc WNDPROC = new WindowProc() { 
+        @Override
+        public LRESULT callback(HWND hWnd, int uMsg, WPARAM wParam, LPARAM lParam)
+        {
+            if (uMsg == WM_COPYDATA) {
+
+                COPYDATASTRUCT copydatastruct = new COPYDATASTRUCT(new Pointer(lParam.longValue()));
+
+                /* Split message data at zero bytes. */
+                String[] arr = copydatastruct.lpData.getWideString(0).split("\\\\0");
+                
+                /* End the process if the message data is a duplicate of the previous. */ 
+                if (Arrays.deepEquals(arr, _msnArray)) { return new LRESULT(0); };
```

Comment:
Duplicate handling shouldn't be there and on top of that it's handled incorrectly. 
Assume player retrys already played map - request gets ignored.